src/AST-Core-Tests/RBBlockNodeTest.class.st

```diff
@@ -25,6 +25,34 @@ RBBlockNodeTest >> testHasNonLocalReturn [
 	self deny: (Object>>#halt) ast hasNonLocalReturn
 ]
 
+{ #category : #tests }
+RBBlockNodeTest >> testIsClean [
+	| escpWrite escpRead |
+	escpRead := escpWrite := 1.
+	self deny: [ self yourself ] sourceNode isClean.
+	self deny: [ ^ 1 ] sourceNode isClean.
+	self deny: [ testSelector foo ] sourceNode isClean.
+	self deny: [ escpRead foo ] sourceNode isClean.
+	self deny: [ escpWrite := 2 ] sourceNode isClean.
+	 
+	self deny: [[ self yourself ]] sourceNode isClean.
+	self deny: [[ ^ 1 ]] sourceNode isClean.
+	self deny: [[ testSelector foo ]] sourceNode isClean.
+	self deny: [[ escpRead foo ]] sourceNode isClean.
+	self deny: [[ escpWrite := 2 ]] sourceNode isClean.
+	self assert: [  ] sourceNode isClean.
+	self assert: [ thisContext ] sourceNode isClean.
+	self assert: [ 1 + 2 ] sourceNode isClean.
+	self assert: [ :a | a + 2 ] sourceNode isClean.
+	self assert: [ :a :b | a + b + 3 ] sourceNode isClean.
+	self assert: [ | a | a := 1. a + 3 ] sourceNode isClean.
+	self assert: [ true ifTrue: [  ] ]  sourceNode isClean.
+	"optimized blocks"
+	self assert: [ 
+		| local2 | true ifTrue: [local2 := 1 ] ] isClean. "optimized blocks are clean"
+	self deny: [ true ifTrue: [escpWrite := 1 ] ] isClean. "But not if we have a real escape"
+]
```

Comment:
shouldn't these tests be shared with BlockClosureTests?

Reply:
There is 

[ ] isClean
and
[ ] sourceNode isClean

For now this is the same, but later #isClean on BlockClosure hierachy will not delegate to the AST.

Maybe having the same code with just  "sourceNode" removed and a comment in both tests to keep them in sync would be an idea

Reply:
(the #sourceNode send can. not be in the Kernel tests as they are carefully structured to not need the AST/Compiler)